src/Python/divik2spectre.py

```diff
@@ -0,0 +1,149 @@
+        self.data = data
+        self.xy = xy
+        self.mz = mz.reshape((-1,))
+        self.region = region.reshape((-1))
+        self.region_names = region_names
+        self._serialized = None
+
+    def __str__(self):
+        if self._serialized is None:
+            header = '\n'.join([
+                'region names: ' + ', '.join(self.region_names),
+                ' '.join(map(str, self.mz))
+            ])
+            self._serialized = '\n'.join([
+                header,
+                *[_serialize_entry(spectrum, location, region)
+                  for spectrum, location, region
+                  in zip(tqdm(self.data, desc='Spectrum'), self.xy, self.region)]
+            ])
+        return self._serialized
+
+
+def _parse_hdf5_cell_array(object_reference, matfile) -> List[str]:
+    """Parses MATLAB cell array of character arrays from hdf5 file
+
+    :param object_reference: cell array in the file
+    :param matfile: mat-file parsed
+    :return: list of strings
+    """
+    data = []
+    for column in object_reference:
+        row_data = []
+        for row_number in range(len(column)):
+            row_data.append(''.join(map(chr, matfile[column[row_number]][:])))
+        data.append(row_data)
+    data = list(np.transpose(data)[0])
+    return data
+
+
+def read_hdf5(path: str) -> DivikDataSerializer:
+    """Parses hdf5 mat-file with DiviK data into serializer"""
+    matfile = h5py.File(path, 'r')
+    region_names = _parse_hdf5_cell_array(matfile['region_names'], matfile)
+    data = np.transpose(matfile['data'])
+    xy = np.transpose(matfile['xy']).astype(int)
+    assert data.shape[0] == xy.shape[0]
+    assert xy.shape[1] == 2
+    region = np.transpose(matfile['region']).astype(int)
+    assert region.size == data.shape[0]
+    if 'mz' not in matfile:
+        print('M/Z array not found, creating fake...')
+        mz = np.arange(data.shape[1])
+    else:
+        mz = np.transpose(matfile['mz'])
+        assert mz.size == data.shape[1]
+    return DivikDataSerializer(data, xy, mz, region, region_names)
+
+
+def read_matfile(path: str) -> DivikDataSerializer:
+    """Parses v7.0 >= mat-file with DiviK data into serializer"""
+    matfile = scio.loadmat(path)
+    region_names = [name_table[0] for name_table in matfile['region_names'][0]]
+    data = matfile['data']
+    xy = matfile['xy']
+    assert data.shape[0] == xy.shape[0]
+    assert xy.shape[1] == 2
+    region = matfile['region']
+    assert region.size == data.shape[0]
+    if not 'mz' in matfile:
+        print('M/Z array not found, creating fake...')
+        mz = np.arange(data.shape[1])
+    else:
+        mz = matfile['mz']
+        assert mz.size == data.shape[1]
+    return DivikDataSerializer(data, xy, mz, region, region_names)
+
+
+def main():
+    args = parse_args()
+    try:
+        data = read_matfile(args.source)
+    except NotImplementedError:
+        data = read_hdf5(args.source)
```

Comment:
Just wondering, where can `read_matfile` raise a `NotImplementedError`?

Reply:
It raises when we're trying to read a `-v7.3` file with `scipy` reader - it does not implement `hdf5` files. This way we have automated version detection instead of forcing user to pass this information.

Reply:
Sure thing. Thanks for clarification!

Reply:
n1